src/grpc/infer_handler.cc

```diff
@@ -1053,21 +1046,31 @@ ModelInferHandler::InferResponseComplete(
     response->Clear();
   }
 
-  ::grpc::Status status;
-  GrpcStatusUtil::Create(&status, err);
+  GrpcStatusUtil::Create(&state->status_, err);
   TRITONSERVER_ErrorDelete(err);
 
   LOG_TRITONSERVER_ERROR(
       TRITONSERVER_InferenceResponseDelete(iresponse),
       "deleting GRPC inference response");
 
+  // Defer sending the response until FINAL flag is seen or
+  // there is error
+  if ((flags & TRITONSERVER_RESPONSE_COMPLETE_FINAL) == 0) {
+    return;
+  }
+
+#ifdef TRITON_ENABLE_TRACING
+  state->trace_timestamps_.emplace_back(std::make_pair(
+      "INFER_RESPONSE_COMPLETE", TraceManager::CaptureTimestamp()));
+#endif  // TRITON_ENABLE_TRACING
+
 #ifdef TRITON_ENABLE_TRACING
   state->trace_timestamps_.emplace_back(
       std::make_pair("GRPC_SEND_START", TraceManager::CaptureTimestamp()));
 #endif  // TRITON_ENABLE_TRACING
```

Comment:
```suggestion
#ifdef TRITON_ENABLE_TRACING
  state->trace_timestamps_.emplace_back(std::make_pair(
      "INFER_RESPONSE_COMPLETE", TraceManager::CaptureTimestamp()));
  state->trace_timestamps_.emplace_back(
      std::make_pair("GRPC_SEND_START", TraceManager::CaptureTimestamp()));
#endif  // TRITON_ENABLE_TRACING
```

Reply:
But we might want to capture `"INFER_RESPONSE_COMPLETE"`  where it was in the code before

Reply:
Perhaps, I am not sure as I just notice that the timing of the `RESPONSE_COMPLETE` capture was different in GRPC v.s. in HTTP. We can either keep the timing in previous implementation or unify it in both endpoints.

In HTTP, it is captured after the HTTP response has been fully prepared and right before Triton invoke the HTTP library to send the response. Whereas in GRPC, the timestamp is captured when the callback is invoked (before preparing the GRPC response).

Reply:
resolved offline